src/integration/resources/inprocess/cluster.go

```
@@ -0,0 +1,314 @@
+
+	// Ensure that once we start creating resources, they all get cleaned up even if the function
+	// fails half way.
+	defer func() {
+		if err != nil {
+			cleanup(logger, nodes, coord)
+		}
+	}()
+
+	for i := 0; i < numNodes; i++ {
+		cfg := defaultDBNodesCfg
+		dbnodeOpts := defaultDBNodeOpts
+
+		if i == 0 {
+			// Mark the initial node as the etcd seed node.
+			dbnodeOpts.GenerateHostID = false
+			cfg.DB.HostID = &hostid.Configuration{
+				Resolver: hostid.ConfigResolver,
+				Value:    &hostID,
+			}
+		}
+		cfg.DB.Discovery = &discoveryCfg
+
+		var node resources.Node
+		node, err = NewDBNode(cfg, dbnodeOpts)
+		if err != nil {
+			return nil, err
+		}
+		nodes = append(nodes, node)
+	}
+
+	var coordConfig coordinatorcfg.Configuration
+	coordConfig, err = opts.Coordinator.Config.ToConfig()
+	if err != nil {
+		return nil, err
+	}
+	// TODO(nate): refactor to support having envconfig if no DB.
+	coordConfig.Clusters[0].Client.EnvironmentConfig = &envConfig
+	coord, err = NewCoordinator(coordConfig, CoordinatorOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	m3 := NewM3Resources(ResourceOptions{
+		Coordinator: coord,
+		DBNodes:     nodes,
+	})
+	if err = resources.SetupCluster(m3, &resources.ClusterOptions{
+		ReplicationFactor: int32(opts.DBNode.RF),
+		NumShards:         int32(opts.DBNode.NumShards),
+	}); err != nil {
+		return nil, err
+	}
+
+	return m3, nil
+}
+
+// generateDefaultDiscoveryConfig handles creating the correct config
+// for having an embedded ETCD server with the correct server and
+// client configuration.
+func generateDefaultDiscoveryConfig(
+	cfg dbcfg.Configuration,
+	hostID string,
+) (discovery.Configuration, environment.Configuration, error) {
+	discoveryConfig := cfg.DB.DiscoveryOrDefault()
+	envConfig, err := discoveryConfig.EnvironmentConfig(hostID)
+	if err != nil {
+		return discovery.Configuration{}, environment.Configuration{}, nil
```

error is not returned here?